API-4440 - API Publishing Expectations page

alexandervalencia opened the pull request:
### Description

This PR supports [API-4440](https://vajira.max.gov/browse/API-4440) by adding an API Expectations page.

Because there was so much content, I opted to use markdown files (`.mdx`). In regards to testing that the markdown is being correctly loaded: our Jest configuration does not currently handle loading markdown. I explored trying to make that a reality, but unfortunately, with packages that are currently available (`jest-mdx-loader` and `jest-transformer-mdx` were the two I tested), it's not possible to load the markdown files for testing—at least not possible in a "non-hacky" way.

At some point in the future if we want to test markdown content properly across the entire developer-portal, I would recommend looking into adapting the custom fork we have of the `markdown-component-loader` to be adapted to Jest.
<!--
What is your PR doing and why? Please link a ticket and any other relevant
relations like Slack threads or Sentry issues.

If you're making non-content updates make sure you have read the development guide: https://github.com/department-of-veterans-affairs/developer-portal/blob/master/docs/development.md
-->

### Process

<!--
All new components should have associated unit tests, at minimum. If your PR is modifying any
existing component with little or no testing can you improve the testing in that corner of the codebase?

Is your change something that needs an update in the documentation? (Referring specifically to documentation
for operating the developer portal and not documentation for that APIs that live on the developer portal).

If neither of these checks are relevant to your PR (like changes that only affect content) you can delete them.
-->

- [X] Tests added or updated for change
- [ ] Docs updated
  - [ ] If you have made structural changes to the site, the [Developer Portal Content Types](https://community.max.gov/x/qUwOg) docs have been updated
  - [ ] If you have made/confirmed any decisions about user interactions and accessibility, the [Accessible Component Design](https://community.max.gov/x/xS8mgg) docs have been updated
- [ ] Accessibility concerns have been considered and addressed
  <!-- Deque's axe browser extension: https://www.deque.com/axe/browser-extensions/ -->
  - [ ] If you've made any UI changes, you've run an axe check using the axe browser extension that reported no new issues

### Requested feedback

<!-- Is there any specific feedback you are looking for on the PR? It's okay if this is blank. -->

mjlumetta commented on src/containers/publishing/components/publishingExpecations/PublishingExpectations.test.tsx:
```diff
@@ -0,0 +1,71 @@
+  it('renders all accordions in a closed state', () => {
+    const buttons = screen.getAllByRole('button');
+    buttons.forEach(button => {
+      expect(button.getAttribute('aria-expanded')).toBe('false');
```
nice test!!

alexandervalencia replied:
Thanks!

mjlumetta commented on src/containers/publishing/components/publishingExpecations/PublishingExpectations.tsx:
```diff
@@ -0,0 +1,77 @@
+
+const PublishingExpectations = (): JSX.Element => {
+  React.useEffect(() => {
+    // CollapsiblePanel expects a VetsGov object on the global window
```
hate this so much!!! 🤮 

but actually, if I'm being practical about it... should we add this effect in `App` or somewhere else at the top level and then forego the individual declarations everywhere we need `CollapsiblePanel`?

alexandervalencia replied:
me too 😑  

Yeah, I think that's a good idea since we know it's going to be used in at least two spots.

mjlumetta commented on src/content/publishing/expectations/security.mdx:
```diff
@@ -0,0 +1,49 @@
+
+Lighthouse has a full authorization to operate (ATO) at the FISMA Moderate level. We also complete a privacy impact assessment and privacy threshold analysis whenever the API gateway sends new kinds of data to new systems.
+
+The production environment is [accessible here](https://api.va.gov) and the sandbox environment is [accessible here](https://dev-api.va.gov). Our production and sandbox environments are deployed in the VA Enterprise Cloud's AWS GovCloud region, sitting behind the VA Trusted Internet Connection (TIC), and in compliance with VA’s external system network requirements.
```
```suggestion
The production environment is [accessible here](https://api.va.gov) and the sandbox environment is [accessible here](https://sandbox-api.va.gov). Our production and sandbox environments are deployed in the VA Enterprise Cloud's AWS GovCloud region, sitting behind the VA Trusted Internet Connection (TIC), and in compliance with VA’s external system network requirements.
```

mjlumetta replied:
also gonna drop this issue Brent flagged here https://lighthouseva.slack.com/archives/C0190E2PVTQ/p1611958892171700?thread_ts=1611953150.159900&cid=C0190E2PVTQ (axe will raise it as well)

mjlumetta commented on src/content/publishing/expectations/security.mdx:
```diff
@@ -0,0 +1,49 @@
+
+### Addressing Security Risks
+
+If you have a simple API that publishes shared information about VA, there are very few security risks to address, and we use a variety of methods to provide authentication and security. You can read more about this on our [Authentication and Authorization](https://developer.va.gov/explore/verification/docs/authorization) page, but to start, here are some of the ways our authentication offers add ...
```
the Authorization link should be an internal link using React Router. technically this works, but there are two considerations here:

1. with an absolute href, it doesn't respect the host in lower environments.
2. if it's not a `Link` or a `HashLink`, it will reload the whole site when clicked. even though this is probably not the end of the world, especially if the site is cached in the user's browser, it's not consistent with the client-side routing experience, and we should probably keep the single-page experience consistent.

also, should we use the v2 auth docs flag to conditionally show the old/new URLs? this link should be to `/explore/authorization` in the near future.

alexandervalencia replied:
Great observations, I added the flag and made sure to use `Link`

mjlumetta commented on src/content/publishing/expectations/authentication.mdx:
```diff
@@ -0,0 +1,52 @@
+
+API keys are maintained within Kong. There is no default expiration date. If a key needs to be rotated on a recurring basis, Lighthouse can either update the key immediately, or update it on a rolling basis, where both keys work for a short period of time and the consumer facilitates the update on their end with no downtime. The information passed to the API provider does not change when the key  ...
+Sandbox and production access
+API consumers are granted access to the sandbox environment first, so they can test your API without data risk. When they’re ready, they become approved for production access using our [Path to Production](https://github.com/department-of-veterans-affairs/vets-api-clients/blob/master/Path-to-Production.md).
```
this link points the user back to the `/go-live` page. should this link directly to that page instead?

brentschneider replied:
That would make the most sense to me as well Mike.  Let's link directly to the portal `/go-live` page.

alexandervalencia replied:
confirmed with Brent 👍

mjlumetta commented on src/content/publishing/expectations/authentication.mdx:
```diff
@@ -0,0 +1,52 @@
+<div className="feature">
+
+### What we provide
```
issue flagged by axe: `markdown-component-loader` generates ids for heading level (think this is due to a `markdown-it` plugin, forget the exact mechanics but can check on this) based on the text content. anyway, the lovely consistent structure of this feature div means that these headings all have duplicate ids. would you mind giving them custom ids using the `{#my-custom-id-name}` annotation? you can see examples of this on the Release Notes page, where we've run into this issue before with dates (particularly with OAuth APIs), and on the new Auth docs components.

mjlumetta commented on src/content/publishing/expectations/authentication.mdx:
```diff
@@ -0,0 +1,52 @@
+- **SSOe**: single sign on for external users, such as Veterans and associates signing into public-facing websites like VA.gov, or using a third-party external app integrated with VA data
+
+API keys are maintained within Kong. There is no default expiration date. If a key needs to be rotated on a recurring basis, Lighthouse can either update the key immediately, or update it on a rolling basis, where both keys work for a short period of time and the consumer facilitates the update on their end with no downtime. The information passed to the API provider does not change when the key  ...
+Sandbox and production access
```
what's going on here - is this supposed to be a heading? looks a bit wonky when rendered.

alexandervalencia replied:
Yeah, missed a heading here. Good catch! 🧤

mjlumetta commented on src/content/publishing/expectations/authentication.mdx:
```diff
@@ -0,0 +1,52 @@
+<div className="feature">
```
for these feature divs, there's no equivalent of the visual highlighting for screen readers. should we give them some kind of markup that will distinguish them as such? my initial thought would be a `role="region"` + an `aria-label`, but I'm open to suggestion.

alexandervalencia replied:
`region` seems like the correct role here, I think you're spot on!
ref: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Region_role

mjlumetta reviewed (commented):
nice job on this! comments are mostly accessibility concerns + copy edits.

mjlumetta commented:
re: the Jest/Markdown question, that's a good concern to point out. I've got a couple quick thoughts before signing off for the week:

1. do we want to store this much content in MDX files in the first place? thinking that if we're ever storing and editing content externally in a consistent way, that may cut out the need for this conversation entirely.
2. if there's anything that urgently needs testing, integration tests are always an option. 

but yeah, would love to solve this at some point.

alexandervalencia commented:
Thanks for reviewing @mjlumetta! I've hopefully addressed all the concerns you mentioned. In regards to your final thoughts on MDX. I'm with you, it would be much nicer if we could store this much content elsewhere in a more easily editable way for non-devs to handle. Until then, I think having mostly pure-content is better than large `tsx` files.

mjlumetta commented on src/containers/publishing/components/publishingExpecations/PublishingExpectations.test.tsx:
```diff
@@ -0,0 +1,71 @@
+    });
+  });
+
+  describe('security', () => {
```
absolutely not necessary, but if you end up pushing another commit and are interested, might be a good use case for `describe.each`. that's purely aesthetic though

mjlumetta commented on src/types/md.d.ts:
```diff
@@ -7,6 +7,7 @@ declare module '*.mdx' {
    * preference to limit our use the component loader to its basic functionality.
    */
   interface MarkdownComponentProps {
+    [x: string]: unknown;
```
couple small notes here: might want to check with @ediiotero on this one, since he's got similar but differently structured changes on #611. also, regardless of how we commit this functionality, the comment above is no longer accurate, so that will need to change, I think.

gileswells replied:
I personally like this method better than what's in #611. I'm adding a review comment over on that ticket about this.

mjlumetta commented on src/content/publishing/expectations/versioning.mdx:
```diff
@@ -0,0 +1,41 @@
+
+Breaking changes to published API versions can be disruptive to consumers, so our policy is to ensure breaking changes aren’t published without a major version change.
+
+We define a breaking change as any change that might adversely impact a consumer’s use of an API. For example, removing an endpoint or field would be classified as a breaking change. A non-breaking change is one that does not adversely impact current users of an API, such as adding an endpoint or field. Non-breaking changes do not require a new version, but changes that could impact consumers’ ex ...
```
one more that should be a `Link`

alexandervalencia replied:
Thanks for the catch!

mjlumetta commented on src/content/publishing/expectations/authentication.mdx:
```diff
@@ -0,0 +1,59 @@
+  '{ Link }': 'react-router-dom'
+---
+
+<div className="feature" role="region" aria-label="Authentication and Authorization Expectations">
```
the region + ARIA label thing works great! only thing I'd ask here is should we tweak the labels to be "Authentication and Authorization Expectations Summary" (or "overview")? since the whole page is about expectations, might be important to use language that clarifies that these are the key points and alludes to additional info below. (especially in case someone is navigating by landmarks, don't want them to think the region is the whole body of information we're providing.)

alexandervalencia replied:
love this, will fix!

mjlumetta reviewed (commented):
yeah, this is basically there. I've suggested a small language tweak on the ARIA labels for the feature regions, there's a missing `Link` element, and would suggest syncing up with @ediiotero on the Markdown props type def since he's working on something similar. but overall looks great, this is nearly ready to go.

mjlumetta reviewed (dismissed):
lgtm! good to go pending CI checks ✨

rtravitz commented:
@mjlumetta Any thoughts on why that lint comment triple posted?

alexandervalencia commented:
Brent found two more issues; the changes are fixed, it should be okay from him, just need final approval here, Thanks!

mjlumetta reviewed (approved):
re-approving ✅
